lib/sqlalchemy/sql/schema.py

```diff
@@ -385,6 +385,11 @@ def foreign_keys(self) -> Set[ForeignKey]:
 
     """
 
+    periods: PeriodCollection
+    """A collection of all :class:`_schema.Period` objects associated with this
+        :class:`_schema.Table`.
```

Comment:
**mike bayer** (zzzeek) wrote:

I appreciate the PeriodCollection namespace but I think using this here is not consistent vs. how all the other elements on Table are stored, outside of Column objects.  Column objects are in the .c namespace because they are integral to the SQL expression API, where access to the columns on a table is ubiquitous throughout code.  however, access to DDL-only structures like indexes, foreign keys, primary keys, etc. is infrequent, so we have simpler collections for these things like sets and dictionaries.

all that is to say, I'd prefer that "periods" be a built in collection type, here it seems like a Python dictionary would be most expedient.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803

Reply:
**Trevor Gross** (tgross35) wrote:

Understood, that makes sense. I started with just a set but switched because of thinking about DML - the application period constructions like `WHERE eperiod CONTAINS DATE '2011-01-02'` would map nicely to something like `.where(table.periods.eperiod.contains('2011-01-02')`. 

Bit of an aside - of course, it could be simplified further. Since only one application time period is allowed, that could become a table attribute that works like `.where(table.application_period.contains(...)` regardless of how the period is named. The only reason I didn't go for this is because while the spec technically says one SV and one AV period per table, it is definitely designed so that a backend could have any number AV periods if they wanted. So I kind of worked with the goal of closely following what the ISO SQL DDL/DML looks like rather than how it's currently implemented.

All that being said, if you'd maybe be OK with a subclass of dict with __getattr__ and __getitem__ to allow dot access (to work with the where clause example above), I think that would get rid of the collection overkill but keep the flexibility for whenever the DML is ready.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803

Reply:
**Federico Caselli** (CaselIT) wrote:

periods seems to be somewhat similar to some kind of system columns (can appear in constraint, in the where, etc).

Should we treat them like some kind of columns instead, like a Column(system=True, period=True) or similar? It would not be added as a column but it would install itself as such.

This is just a though, posting to see opinions on this

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803

Reply:
**Federico Caselli** (CaselIT) wrote:

trying a bit, and no, it does not seem a column.

not sure what db support `WHERE eperiod CONTAINS DATE '2011-01-02'` but mariadb does not seem to like it, and from looking at the doc neither mssql. @Trevor where did you find that syntax?

DQL seems to be mostly related to a table, so in a future pr a method of table like `for_period(name)` that could return a kind of table_valued object that could be used with `.asof(...)`/`.between(...)`/... to render something like
 
  FROM t FOR SYSTEM_TIME BETWEEN (NOW() - INTERVAL 1 YEAR) AND NOW();
  FROM t FOR SYSTEM_TIME AS OF TIMESTAMP'2016-10-09 08:07:06';

given this usage I'm not sure we need the .period accessor to be anything special

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803